File: ua8295_emulator/memory.py
from __future__ import annotations

from dataclasses import dataclass, field


def _fit64k(data: bytes, fill: int = 0xFF) -> bytearray:
    if len(data) > 0x10000:
        raise ValueError("memory image is larger than 64 KB")
    return bytearray(data + bytes([fill]) * (0x10000 - len(data)))


@dataclass
class ExternalBus:
    """Simple 80C31 external code/data bus with traceable peripheral fallbacks."""

    code: bytes
    xram_size: int = 0x10000
    default_read: int = 0xFF
    text_rom: bytes = b""
    io_events: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        self.code_mem = _fit64k(self.code)
        if not (1 <= self.xram_size <= 0x10000):
            raise ValueError("xram_size must be between 1 and 65536")
        self.xram = bytearray([0x00] * self.xram_size)
        self.text_rom_mem = bytes(self.text_rom)

    def read_code(self, addr: int) -> int:
        return self.code_mem[addr & 0xFFFF]

    def read_xdata(self, addr: int) -> int:
        addr &= 0xFFFF
        if addr < len(self.xram):
            return self.xram[addr]
        if self.text_rom_mem:
            text_addr = addr - 0x8000
            if 0 <= text_addr < len(self.text_rom_mem):
                return self.text_rom_mem[text_addr]
        self.io_events.append(f"read_xdata 0x{addr:04X} -> 0x{self.default_read:02X}")
        return self.default_read

    def write_xdata(self, addr: int, value: int) -> None:
        addr &= 0xFFFF
        value &= 0xFF
        if addr < len(self.xram):
            self.xram[addr] = value
            return
        self.io_events.append(f"write_xdata 0x{addr:04X} <- 0x{value:02X}")

    def read_pdata(self, page: int, low_addr: int) -> int:
        return self.read_xdata(((page & 0xFF) << 8) | (low_addr & 0xFF))

    def write_pdata(self, page: int, low_addr: int, value: int) -> None:
        self.write_xdata(((page & 0xFF) << 8) | (low_addr & 0xFF), value)

File: src/memory.ts
export function fit64k(data: Uint8Array, fill = 0xff): Uint8Array {
  if (data.length > 0x10000) {
    throw new Error("memory image is larger than 64 KB");
  }
  const out = new Uint8Array(0x10000);
  out.fill(fill);
  out.set(data);
  return out;
}

export type XdataRegion = "xram" | "text-rom" | "unmapped";

export class ExternalBus {
  readonly codeMem: Uint8Array;
  readonly xram: Uint8Array;
  readonly textRom: Uint8Array;
  readonly ioEvents: string[] = [];
  private readonly xramBase: number;

  constructor(
    code: Uint8Array,
    options: { xramBase?: number; xramSize?: number; defaultRead?: number; textRom?: Uint8Array } = {}
  ) {
    const xramSize = options.xramSize ?? 0x10000;
    if (xramSize <= 0 || xramSize > 0x10000) {
      throw new Error("xramSize must be between 1 and 65536");
    }
    this.xramBase = options.xramBase ?? 0;
    this.codeMem = fit64k(code);
    this.xram = new Uint8Array(xramSize);
    // Mimic real CMOS SRAM / EPROM power-up state: bytes default to 0xFF rather
    // than the Uint8Array convention of 0x00. The firmware uses 0xFF as the
    // "empty slot" marker for the message buffer at 0x6800+ (lookup routine at
    // main code 0x116F reads MOVX, XRLs with #0xFF, then JZ-branches on
    // empty). Initialising XRAM to 0x00 incorrectly told the firmware every
    // slot was allocated, which surfaced as a "MEMORY FULL!" prompt during
    // message entry on the device-mode workflow probe.
    this.xram.fill(0xff);
    this.textRom = options.textRom ?? new Uint8Array();
    this.defaultRead = options.defaultRead ?? 0xff;
  }

  private readonly defaultRead: number;

  readCode(addr: number): number {
    return this.codeMem[addr & 0xffff];
  }

  readXdata(addr: number): number {
    const masked = addr & 0xffff;
    const xramOffset = masked - this.xramBase;
    if (xramOffset >= 0 && xramOffset < this.xram.length) {
      return this.xram[xramOffset];
    }
    const textAddr = masked - 0x8000;
    if (textAddr >= 0 && textAddr < this.textRom.length) {
      return this.textRom[textAddr];
    }
    this.ioEvents.push(`read_xdata 0x${hex(masked, 4)} -> 0x${hex(this.defaultRead, 2)}`);
    return this.defaultRead;
  }

  regionForXdata(addr: number): XdataRegion {
    const masked = addr & 0xffff;
    const xramOffset = masked - this.xramBase;
    if (xramOffset >= 0 && xramOffset < this.xram.length) {
      return "xram";
    }
    const textAddr = masked - 0x8000;
    if (textAddr >= 0 && textAddr < this.textRom.length) {
      return "text-rom";
    }
    return "unmapped";
  }

  writeXdata(addr: number, value: number): void {
    const masked = addr & 0xffff;
    const byte = value & 0xff;
    const xramOffset = masked - this.xramBase;
    if (xramOffset >= 0 && xramOffset < this.xram.length) {
      this.xram[xramOffset] = byte;
      return;
    }
    this.ioEvents.push(`write_xdata 0x${hex(masked, 4)} <- 0x${hex(byte, 2)}`);
  }

  readPdata(page: number, lowAddr: number): number {
    return this.readXdata(((page & 0xff) << 8) | (lowAddr & 0xff));
  }

  writePdata(page: number, lowAddr: number, value: number): void {
    this.writeXdata(((page & 0xff) << 8) | (lowAddr & 0xff), value);
  }
}

export function hex(value: number, width: number): string {
  return (value >>> 0).toString(16).toUpperCase().padStart(width, "0");
}
